src/coreclr/System.Private.CoreLib/src/System/Array.CoreCLR.cs

```diff
@@ -554,8 +413,94 @@ private unsafe nint GetFlattenedIndex(ReadOnlySpan<int> indices)
+            }
+
+            Debug.Assert((nuint)flattenedIndex < NativeLength);
+
+            ref byte arrayDataRef = ref MemoryMarshal.GetArrayDataReference(this);
+
+            MethodTable* pElementMethodTable = arrayElementTypeHandle.AsMethodTable();
+
+            if (value == null)
+            {
+                // Null is the universal zero...
+                if (pElementMethodTable->IsValueType)
+                {
+                    ref byte offsetDataRef = ref Unsafe.Add(ref arrayDataRef, flattenedIndex * pMethodTable->ComponentSize);
+                    nuint elementSize = pElementMethodTable->GetNumInstanceFieldBytes();
+                    if (pElementMethodTable->ContainsGCPointers)
+                        SpanHelpers.ClearWithReferences(ref Unsafe.As<byte, nint>(ref offsetDataRef), elementSize / (nuint)sizeof(IntPtr));
+                    else
+                        SpanHelpers.ClearWithoutReferences(ref offsetDataRef, elementSize);
+                }
+                else
+                {
+                    Unsafe.Add(ref Unsafe.As<byte, object?>(ref arrayDataRef), (nuint)flattenedIndex) = null;
+                }
+            }
+            else if (!pElementMethodTable->IsValueType)
+            {
+                if (pElementMethodTable != TypeHandle.TypeHandleOf<object>().AsMethodTable() //  Everything is compatible with Object
+                    && CastHelpers.IsInstanceOfAny(pElementMethodTable, value) == null)
+                    throw new InvalidCastException(SR.InvalidCast_StoreArrayElement);
+
+                Unsafe.Add(ref Unsafe.As<byte, object?>(ref arrayDataRef), (nuint)flattenedIndex) = value;
+            }
+            else
+            {
+                // value class or primitive type
+
+                ref byte offsetDataRef = ref Unsafe.Add(ref arrayDataRef, flattenedIndex * pMethodTable->ComponentSize);
+                if (CastHelpers.IsInstanceOfAny(pElementMethodTable, value) != null)
+                {
+                    if (pElementMethodTable->IsNullable)
+                    {
+                        RuntimeHelpers.Unbox_Nullable(ref offsetDataRef, pElementMethodTable, value);
+                    }
+                    else
+                    {
+                        nuint elementSize = pElementMethodTable->GetNumInstanceFieldBytes();
+                        if (pElementMethodTable->ContainsGCPointers)
+                        {
+                            Buffer.BulkMoveWithWriteBarrier(ref offsetDataRef, ref value.GetRawData(), elementSize);
+                        }
+                        else
+                        {
+                            SpanHelpers.Memmove(ref offsetDataRef, ref value.GetRawData(), elementSize);
+                        }
+                    }
+                }
+                else
+                {
+                    // Allow enum -> primitive conversion, disallow primitive -> enum conversion
```

Comment:
The enum->primitive conversions are not handled correctly after this change and we are missing tests to catch that. Try:
```
var a = new int[1];
a.SetValue(DayOfWeek.Monday, 0);
Console.WriteLine(a[0]);
```
Result:
```
Process terminated. Assertion failed.
Array.Copy from I4 to another type hit unsupported widening conversion
   at System.Array.InternalSetValue(Object value, IntPtr flattenedIndex)
   at Program.<Main>$(String[] args)
```